core/src/test/java/nu/yona/server/util/AsyncExecutorTest.java

```diff
@@ -0,0 +1,131 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotEquals;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.CountDownLatch;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.slf4j.MDC;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.ComponentScan;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.FilterType;
+import org.springframework.scheduling.annotation.EnableAsync;
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.context.junit.jupiter.SpringExtension;
+
+import nu.yona.server.entities.UserRepositoriesConfiguration;
+import nu.yona.server.rest.PassThroughHeadersHolder;
+
+@Configuration
+@EnableAsync
+@ComponentScan(useDefaultFilters = false, basePackages = { "nu.yona.server.rest", "nu.yona.server.util" }, includeFilters = {
+		@ComponentScan.Filter(pattern = "nu.yona.server.rest.PassThroughHeadersHolder", type = FilterType.REGEX),
+		@ComponentScan.Filter(pattern = "nu.yona.server.util.AsyncExecutor", type = FilterType.REGEX) })
+class AsyncExecutorTestConfiguration extends UserRepositoriesConfiguration
+{
+}
+
+@ExtendWith(SpringExtension.class)
+@ContextConfiguration(classes = { AsyncExecutorTestConfiguration.class })
+class AsyncExecutorTest
+{
+	@Autowired
+	private AsyncExecutor service;
+
+	@Autowired
+	private PassThroughHeadersHolder headersHolder;
+
+	@Test
+	void getThreadData_successful_dataIsPresent() throws InterruptedException
+	{
+		DataCarrier expectedData = initializeThread();
+		CountDownLatch doneSignal = new CountDownLatch(1);
+		DataCarrier actionHandlerData = new DataCarrier();
+		DataCarrier completionHandlerData = new DataCarrier();
+
+		service.execAsync(service.getThreadData(), () -> actionHandler(actionHandlerData, Optional.empty()),
+				(t) -> completionHandler(doneSignal, completionHandlerData, t));
+		doneSignal.await();
+
+		assertReturnedData(completionHandlerData, expectedData);
+		assertReturnedData(actionHandlerData, expectedData);
+	}
+
+	@Test
+	void getThreadData_failed_dataIsPresentThrowableIsAvailable() throws InterruptedException
+	{
+		DataCarrier expectedData = initializeThread();
+		expectedData.exception = Optional.of(new NullPointerException("This fails badly"));
+		CountDownLatch doneSignal = new CountDownLatch(1);
+		DataCarrier actionHandlerData = new DataCarrier();
+		DataCarrier completionHandlerData = new DataCarrier();
+
+		service.execAsync(service.getThreadData(), () -> actionHandler(actionHandlerData, expectedData.exception),
+				(t) -> completionHandler(doneSignal, completionHandlerData, t));
+		doneSignal.await();
+
+		assertReturnedData(completionHandlerData, expectedData);
+		expectedData.exception = Optional.empty();
```

Comment:
I am a bit puzzled by this last assert about action handler data. I think it is easier to create 3 tests instead of one, so that this line is not needed:

1. `execAsync_default_dataIsPresentInActionHandler`
2. `execAsync_completionWithError_dataIsPresentInCompletionHandlerThrowableIsAvailable`
3. `execAsync_completionSuccess_dataIsPresentInCompletionHandlerThrowableIsEmpty`

Reply:
How about extending ``assertReturnedData`` with a parameter to indicate it's asserting action data or assertion data? Or even making two methods for it?

Reply:
I've created two assertion methods. That makes it clear what's being asserted.

Reply:
That makes it better, but the situation where the action runs into an exception does not need an assert on the action data. So I would still favor three methods that assert a single thing.

Reply:
I'don't see why we shouldn't assert the data of the action handler at the time it will fail. That's still valid to check. Sure, we could omit that entire assert, as it's covered by the first test already, but in the second test it needs to have the same values before it fails. Three tests doesn't add anything, as there are only two scenarios:
* The action handler succeeds
* The action handler fails

In either case, the action handler and the completion handler need to have all data and in the failure case, the completion handler also needs to have the exception.